Modified README to include Model Registry component; augmented the Mo…

dmatrix opened the pull request:
These are changes to the model registry documentation and code snippets. In particular, I have augmented the model registry description, concepts, API workflow, and added examples.

All code snippets tests are [here](https://github.com/dmatrix/mlflow-tests/blob/master/py/mlflow/README.md):

dmatrix committed:
modified README to include Model Registry component; augmented the Model Registry Workflow API with code snippets

tomasatdatabricks committed:
[mlflow/R] Update R client to call record logged model  (#2430)

* Update R client to call record logged model metadata with tracking server.

* fix

* Ran roxygen to update R docs.

* Updated R NAMESPACE to address cran check complaint.

* test fix

* fix test

* fix test

* Addressed review comment + removed forgotten debug string from earlier.

aarondav committed:
[UI] Better error messages for model registry (#2456)

RensDimmendaal committed:
typo: algorhitm->algorithm (#2490)

aarondav committed:
Update model registry frontend JS with API changes (#2476)

stbof commented on docs/source/model-registry.rst:
```diff
@@ -16,31 +16,31 @@ staging to production), and annotations.
 
-The MLflow Model Registry is a standalone component that introduces several concepts that
-describe the full lifecycle of an MLflow Model.
+As the fourth standalone component of MLflow, the Model Registry introduces few concepts that describe and facilitate the full lifecycle of an MLflow model.
```
```suggestion
As the fourth standalone component of MLflow, the Model Registry introduces a few concepts that describe and facilitate the full lifecycle of an MLflow model.
```

stbof replied:
Is there a reason why we are highlighting that this is the "fourth" component?

dmatrix replied:
More from having had three components for a while and having MLflow touted as component-based and modularized, thought we would highlight it. The next or 5th pillar might be Model Serving. I

dmatrix replied:
Thinking more about it, "standalone" may be misleading or ambiguous as much as I wish to convey its modularity, and "fourth" I'm not wedded to it. We can drop both, and just rephrase as
"As a new component of MLflow," or just drop the introductory phrase. Let you decide.

stbof replied:
I vote for dropping the phrase.

stbof commented on docs/source/model-registry.rst:
```diff
@@ -16,31 +16,31 @@ staging to production), and annotations.
+    Each distinct model version can be assigned one stage at any given time. MLflow provides predefined stages for common use-cases such as *Staging*, *Production* or *Archived*. You can transition a model version from one stage to another stage.
 
+Annotations and Descriptions
+    You can annotate the top-level model and each version individually using markdown, including description and any relevant information useful for the team such as algorithm descriptions, dataset employed or methodology.
```
```suggestion
    You can annotate the top-level model and each version individually using Markdown, including description and any relevant information useful for the team such as algorithm descriptions, dataset employed or methodology.
```

stbof commented on docs/source/model-registry.rst:
```diff
@@ -16,31 +16,31 @@ staging to production), and annotations.
+Model Registry Workflows
+========================
 
 Before a model can be added to the Model Registry you must log it using the ``log_model`` methods
```
```suggestion
Before you can add a model to the Model Registry you must log it using the ``log_model`` methods
```

stbof commented on docs/source/model-registry.rst:
```diff
@@ -78,47 +78,170 @@ version to one of the other valid stages.
+                artifact_path="sklearn-model",
+                registered_model_name="sk-learn-random-forest-reg-model")
+
+This will log the model as well as register it under the specified name as version 1.
```
```suggestion
This logs the model as well as registers it under the specified name as version 1.
```

stbof commented on docs/source/model-registry.rst:
```diff
@@ -78,47 +78,170 @@ version to one of the other valid stages.
+    ...
+    ...
+
+With hundreds of models, this call can be cumbersome to peruse. A more appropriate call would be to search a specific model name and list its version
```
```suggestion
With hundreds of models, this call can be cumbersome to peruse. A more appropriate call would be to search for a specific model name and list its version
```

stbof commented on docs/source/model-registry.rst:
```diff
@@ -16,31 +16,31 @@ staging to production), and annotations.
 
 Registered Model
-    An MLflow Model registered with the MLflow Model Registry. A registered model has a unique name, contains versions, and other metadata.
+    An MLflow Model can be registered with the  Model Registry. It has a unique name, contains versions, associated transitional stages, model lineage, and other metadata.
```
```suggestion
    An MLflow Model can be registered with the  Model Registry. A registered model has a unique name, contains versions, associated transitional stages, model lineage, and other metadata.
```

stbof replied:
Be careful about replacing nouns with pronouns. In your previous version, it's not clear if "It" refers to the Model Registry or the Model.

stbof commented on README.rst:
```diff
@@ -14,6 +14,7 @@ currently run ML code (e.g. in notebooks, standalone applications or the cloud).
 * `MLflow Models <https://mlflow.org/docs/latest/models.html>`_: A model packaging format and tools that let
   you easily deploy the same model (from any ML library) to batch and real-time scoring on platforms such as
   Docker, Apache Spark, Azure ML and AWS SageMaker.
+* `MLflow Model Registry <https://mlflow.org/docs/latest/model-registry.html>`_: A centralized model store, set of APIs, and UI, to collaboratively manage the full lifecycle of an MLflow Model.
```
```suggestion
* `MLflow Model Registry <https://mlflow.org/docs/latest/model-registry.html>`_: A centralized model store, set of APIs, and UI, to collaboratively manage the full lifecycle of MLflow Models.
```

smurching committed:
Save metric plot state in URLs for easier link sharing (#2393)

Metric plot config is now saved in the URL to simplify sharing plots & collaborating

dmatrix committed:
Update docs/source/model-registry.rst

Co-Authored-By: Stephanie Bodoff <stephanie.bodoff@databricks.com>

dmatrix committed:
Update docs/source/model-registry.rst

+1

Co-Authored-By: Stephanie Bodoff <stephanie.bodoff@databricks.com>

dmatrix committed:
Update docs/source/model-registry.rst

Good idea to have an actor do the task!

Co-Authored-By: Stephanie Bodoff <stephanie.bodoff@databricks.com>

dmatrix committed:
Update docs/source/model-registry.rst

thanks for catching the subject-verb agreement!

Co-Authored-By: Stephanie Bodoff <stephanie.bodoff@databricks.com>

dmatrix committed:
Update docs/source/model-registry.rst

Co-Authored-By: Stephanie Bodoff <stephanie.bodoff@databricks.com>

dmatrix committed:
Update docs/source/model-registry.rst

Co-Authored-By: Stephanie Bodoff <stephanie.bodoff@databricks.com>

dmatrix committed:
Update README.rst

Co-Authored-By: Stephanie Bodoff <stephanie.bodoff@databricks.com>

smurching committed:
Fix bar chart bug introduced in #2393 (#2498)

dmatrix committed:
minor change and incorporated suggestions

dmatrix committed:
Merge branch 'br.model_registry_doc' of github.com:dmatrix/mlflow into br.model_registry_doc

dmatrix committed:
minor typo

t-henri committed:
mlflow gc command (#2265)

* Added hard delete run in sql alchemy store. Added delete artifacts in hdfs store. Added hard delete run endpoint and function in client.

* Fixed delete in hdfs artifact repo. Fixed hard delete endpoint.

* Add support for file store

* Artifacts are deleted on client side when using hard-delete run cli

* Added support for local_artifact_repo

* Fix linter

* Added test for fluent api

* Fix tests

* Added not implemented yet exception for artifact stores other than hdfs and local

* Removed hard delete run endpoint and exposition in client.py

* Added gc command line

* Fix tests

* Remove hard delete runs command line

* Add method _get_deleted_runs in sql alchemy store

* Added test for _get_deleted_runs in sql alchemy store

* Fix linter

* Fix linter

* Added comments to _hard_delete_run and to the gc cli.

* Add implementation of get_deleted_runs for file_store

* delete_artifacts now take a relative path. get_deleted_runs in sql alchemry store returns a list of string. Changes to comments.

* Fix linter

* Fix tests

* Added end to end test for mlflow gc command line

* Fix windows tests

* Fix hdfs_artifact_repo tests on windows

* Fix test_cli for python 2

* Fix linter

* Fix python 2 tests

* Fix python 2 tests

* Retrigeer tests becaused of failed R tests

* Changed urllib import to use six.moves

harupy committed:
Add run links popover to metrics plot (#2295)

* Add runs link popover

* Rename

* Trigger travis

* Use ref and refactor

* Add onRelayout

* Remove unused functions and state

* Refactor

* Add comment

* Fix test

* Fix prop type

* Rename var

* Fix prop type

* Add test for RunLinksPopover

* Add icon

* Refactor

* Close popover when Escaoe key pressed

* use single quotes

* Use named export

* Do not expose state

* Resolve conflict

* Fix popover to use props

* Remove unnecessary comment

* Remove unnecessary comment

* Fix test

* Rename popover title

* Fix style of close button

* Add sorting by y

* Display the popover on the left

* Display metric value

* fix

* Fix

* Fix

* Remove runUuid

* Fix test

* Fix tests

* nit

Commit:
API cleanup (#2502)

* API cleanup

removing `stage` field from `UpdateModelVersion` proto

dbczumar committed:
Minor docs tweaks for MLflow gc command (#2506)

* Added hard delete run in sql alchemy store. Added delete artifacts in hdfs store. Added hard delete run endpoint and function in client.

* Fixed delete in hdfs artifact repo. Fixed hard delete endpoint.

* Add support for file store

* Artifacts are deleted on client side when using hard-delete run cli

* Added support for local_artifact_repo

* Fix linter

* Added test for fluent api

* Fix tests

* Added not implemented yet exception for artifact stores other than hdfs and local

* Removed hard delete run endpoint and exposition in client.py

* Added gc command line

* Fix tests

* Remove hard delete runs command line

* Add method _get_deleted_runs in sql alchemy store

* Added test for _get_deleted_runs in sql alchemy store

* Fix linter

* Fix linter

* Added comments to _hard_delete_run and to the gc cli.

* Add implementation of get_deleted_runs for file_store

* delete_artifacts now take a relative path. get_deleted_runs in sql alchemry store returns a list of string. Changes to comments.

* Fix linter

* Fix tests

* Added end to end test for mlflow gc command line

* Fix windows tests

* Fix hdfs_artifact_repo tests on windows

* Fix test_cli for python 2

* Fix linter

* Fix python 2 tests

* Fix python 2 tests

* Retrigeer tests becaused of failed R tests

* Add/improve documentation

* Fix indent

aarondav commented on docs/source/model-registry.rst:
```diff
@@ -16,31 +16,32 @@ staging to production), and annotations.
 
 Model
-    An MLflow Model logged with one of the model flavor’s ``log_model`` methods.
+    An MLflow Model is created from an experiment or run that is logged with one of the model flavor’s ``mlflow.model_flavor.log_model`` methods. Once logged, this model can then be registered with the Model Registry.
```
Maybe `mlflow.<model_flavor>.log_model` to help clarify that it's not a real snippet that can run?

dmatrix replied:
Yes, indeed, that bit is true. Will change it.

aarondav commented on docs/source/model-registry.rst:
```diff
@@ -78,47 +79,170 @@ version to one of the other valid stages.
+    with mlflow.start_run(run_name="YOUR_RUN_NAME") as run:
+        params={"n_estimators": 5, "random_state": 42}
+        sk_learn_rfr=RandomForestRegressor(params)
+        # Log params using the MLflow sklearn flavor APIs
```
What does this line mean? There is not a flavored param logger.

dmatrix replied:
I meant " Log parameters using the MLflow API"

aarondav commented on docs/source/model-registry.rst:
```diff
@@ -78,47 +79,170 @@ version to one of the other valid stages.
 
 .. code-block:: py
 
-   mlflow.register_model("runs:/e148278107274..832/artifacts/model", "EmailMarketingCampaign")
```
Maybe we can name the arguments like above so it's clear what they are referring to

dmatrix replied:
I think "EmailMarketingCampaign" seems more like an Experiment name you want under which will want to create a set of different types of models—scikit-learn flavor, Keras or PyTorch etc. 

I suppose we can all it WeatherPowerForecast as general if you prefer.

aarondav commented on docs/source/model-registry.rst:
```diff
@@ -78,47 +79,170 @@ version to one of the other valid stages.
 .. code-block:: py
 
-   mlflow.register_model("runs:/e148278107274..832/artifacts/model", "EmailMarketingCampaign")
+    result=mlflow.register_model(“runs:/d16076a3ec534311817565e6527539c0/artifacts/sk-model”,
```
Are these smart quotes, the bane of all copy-pasting?  Maybe make them regular quotes :)

dmatrix replied:
I made them regular quotes, will try do it again. Yes cut-and-paste hast its hidden gotchas.

aarondav commented on docs/source/model-registry.rst:
```diff
@@ -78,47 +79,170 @@ version to one of the other valid stages.
+                source = "mlruns/0/d16076a3ec534311817565e6527539c0/artifacts/sk-model",
+                run_id = "d16076a3ec534311817565e6527539c0")
+
+In contrast, ``mlflow.register_model()`` and ``mlflow.model_flavor.log_model()`` will create a new version in the registry.
```
, if it does not already exist

aarondav commented on docs/source/model-registry.rst:
```diff
@@ -78,47 +79,170 @@ version to one of the other valid stages.
+    client = MlflowClient()
+    client.update_model_version(name=”sk-learn-random-forest-reg-model”,
+            version=1,
+            description="This model version is a random forest containing 100 decision trees that was trained in scikit-learn.")
```
Since the code snippet does not wrap, maybe make the description a bit shorter.

dmatrix replied:
Wil make is shorter

aarondav commented on docs/source/model-registry.rst:
```diff
@@ -78,47 +79,170 @@ version to one of the other valid stages.
+.. code-block:: py
+
+    client=MlflowClient()
+    client.rename_registered_model(name=”sk-learn-random-forest-reg-model”,
```
smart quotes

dmatrix replied:
will fix as smart quotes.

aarondav commented on docs/source/model-registry.rst:
```diff
@@ -78,47 +79,170 @@ version to one of the other valid stages.
+Transitioning an MLflow Model’s Stage
+-------------------------------------
+Over the course of the model’s lifecycle, a model evolves—from development to staging to production.
+You can transition a registered model in the registry to one of the stages: **Staging**, **Production** or **Archived**.
```
None is also possible! You can move a model back to the None stage.

dmatrix replied:
Ah, interesting. Will note that.

mparkhe replied:
Model versions added to registry are in `None` stage. The way this is envisioned is versions moving forward from `Staging` --> `Production` or `Archived` directly. While, the APIs do no block moving versions to `None` it makes little sense. 

Let's **not** document transitioning to `None`. Let's keep it as is.

aarondav commented on docs/source/model-registry.rst:
```diff
@@ -78,47 +79,170 @@ version to one of the other valid stages.
+    client = MlflowClient()
+    client.transition_model_version_stage(name=”sk-learn-random-forest-reg-model”,
+  	        version=3,
+	        stage=“production”)
```
We should probably include an example of `archive_existing_versions`, right @mparkhe ?

dmatrix replied:
I believe this is only an edge feature. We don't expose that in the OSS.

mparkhe replied:
`archive_existing_versions` is not support in SQL backend

aarondav commented on docs/source/model-registry.rst:
```diff
@@ -78,47 +79,170 @@ version to one of the other valid stages.
+
+Deleting Registered Models
+--------------------------
+You can either delete a specific version of the registered model or registered model and all its versions.
```
Clause is a bit confusing, maybe repeat "can delete":

`You can either delete a specific version of the registered model or you can delete the registered model and all its versions.`

dmatrix replied:
Will fix it.

mparkhe replied:
"Archived" is a stage for a model version is a part of lifecycle stage. Usually people would want to keep archived models around for long. Deleting as you mentioned is permanently deleting the model version from the registry. I am not super psyched about mixing these 2 concepts together. How about this.

```
Archiving Models
----------------

To transition model versions out of **Production** stage, transition those model into **Archived** 
stage.

.. code-block:: py

    client = MlflowClient()
    client.transition_model_version_stage(name="sk-learn-random-forest-reg-model",
  	        version=3,
	          stage="Archived")


Deleting Models
---------------

.. note::
  Deleting registered models and model versions is irrevocable, so use it judiciously.

You can either delete a specific version of a registered model or you can delete a registered model 
and all its versions.

.. code-block:: py

    # Delete models versions 1, 2, and 3
    versions = [1, 2, 3]
    client=MlflowClient()
    for version in versions:
      client.delete_model_version(name="sk-learn-random-forest-reg-model", version=version)

    # Delete the registered model along with all its versions
    client.delete_registered_model(name="sk-learn-random-forest-reg-model")

```

aarondav commented on docs/source/model-registry.rst:
```diff
@@ -78,47 +79,170 @@ version to one of the other valid stages.
+    client.delete_registered_model(name=”sk-learn-random-forest-reg-model”)
 
-Newly-registered model versions are added to the registry with stage **None**. To update the stage, call the ``update_model_version`` API:
+However, if you are unsure, you may want to archive specific versions of a registered model.
```
Maybe flip these. We actually expect people to archive most often. Maybe the session could be "Archiving and deleting registered models" and we first say how to archive and then say if you want to permanently delete, use this instead.

dmatrix replied:
Makes sense, better to first Archive them, and later they can always delete once they are certain.

dmatrix committed:
resolved and incorporate suggestions

Commit:
Fixing JS reducer on model version delete (#2507)

mparkhe commented on docs/source/model-registry.rst:
```diff
@@ -78,47 +79,169 @@ version to one of the other valid stages.
+  	        version=3,
+	        stage="production")
+
+Listing and Searching MLflow Registered Models
```
=> Listing and Searching Models

(because `list` an API on registered models and `search` on model versions)

dmatrix replied:
I'm on it. Will commit shortly.

dmatrix replied:
done and addressed it with a new set of commits

mparkhe reviewed (commented):
Thanks @dmatrix. Lemme know if you want me to incorporate these changes.

mparkhe commented on docs/source/model-registry.rst:
```diff
@@ -78,47 +79,169 @@ version to one of the other valid stages.
+       # Log the sklearn model and register as version 1
+       mlflow.sklearn.log_model(sk_model=sk_learn_rfr,
+                artifact_path="sklearn-model",
+                registered_model_name="sk-learn-random-forest-reg-model")
```
This `...` are unrecognized characters.

dmatrix replied:
Ah, no wonder it won't highlight the strings... was wondering.
Will remove it.

dmatrix committed:
resolved and addressed suggestions

Commit:
fixing search test (unrelated to documentation)

Commit:
Merge remote-tracking branch 'jules/br.model_registry_doc' into br.model_registry_doc

mparkhe commented:
@dmatrix : the changes look good. I will push some updates to this PR in order to get the tests to pass so I can get this merged in.

Commit:
fix docs

mparkhe commented:
Failing test is not related to documentation changes. Forcing a merge

harupy commented:
Is this file (added by this PR) necessary?

https://github.com/mlflow/mlflow/blob/master/examples/sklearn_elasticnet_diabetes/osx/ElasticNet-paths.png

mparkhe commented:
> Is this file (added by this PR) necessary?
> 
> https://github.com/mlflow/mlflow/blob/master/examples/sklearn_elasticnet_diabetes/osx/ElasticNet-paths.png

Good catch @harupy -- seems unrelated. I thought it was a missing part of documentation.

@dmatrix: did you intend to link this file to sklearn example? If not -- can you please add a PR to revert this change?

dmatrix commented:
Thanks @mparkhe @harupy. A bit strange that I did not catch this lingering file in my fork. Sorted now: I filed a separate [PR 2493](https://github.com/mlflow/mlflow/pull/2519#) to remove or revert the PNG.
